docs/unit8.mdx

```diff
@@ -0,0 +1,336 @@
+- `COMPSCI161` - This is our parameter, we can change this to another course if we want a different course.
+
+How we know exactly what format to make our URL is usually specified in the API documentation. 
+For the PeterPortal API, we looked here for how to use the courses endpoint. 
+[https://api.peterportal.org/REST-API/courses/](https://api.peterportal.org/REST-API/courses/)
+
+### HTTP Status Codes
+
+In the above response we also received a status code of 200. 
+200 indicates a successful request.
+There are lots of different HTTP status codes, but the gist of it is this:
+
+- 2xx - Success!
+- 4xx - You did something wrong
+- 5xx - There is a problem with the server.
+
+ For more specific status codes, you can read more [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status). 
+
+If we made a request like this:
+
+```json
+cURL https://api.peterportal.org/rest/v0/courses/NONEXISTENT121
+```
+
+Our response will give us an error, indicating this course was not found.
+
+```json
+{
+    "timestamp": "Tue, 15 Feb 2022 22:39:23 GMT",
+    "status": 404,
+    "error": "Bad Request: Invalid parameter",
+    "message": "Course not found"
+}
+```
+
+### URL Encoding
+
+When specifying parameters in URLs, you will need to URL Encode them, especially if there are any special characters. 
+This is because URLs usually use these special characters for other important notation. 
+
+```bash
+# Here the '&' was URL encoded to %26. 
+cURL https://api.peterportal.org/rest/v0/courses/I%26CSCI33
+```
+
+In JavaScript, we can use `encodeURIComponent` to URL encode our parameter strings. 
+
+### Fetch
+
+To make a request to an API in JavaScript, we use the fetch library. 
+
+```bash
+npm install node-fetch
+```
+
+Fetch is a function that takes 2 arguments. The url and a configuration object.
+
+```jsx
+const url = "https://api.peterportal.org/rest/v0/courses/all";
+const response = await fetch(url, {
+	"method": "GET"
+});
+console.log(response.json());
+```
+
+Learn more about Fetch, in Mozilla documentation:
+
+[Using the Fetch API - Web APIs | MDN](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch)
+
+## GraphQL
+
+### What is GraphQL?
+
+💡 GraphQL is a query language for Web APIs. GraphQL itself is *not* a tool or a framework, it is a **language**.
+
+Three cool features of GraphQL
+
+- Get exactly what you ask for
+- Get multiple resources in one request
+- Type System to know what is possible
+
+### Why GraphQL?
+![Social Media Example](../static/img/unit8/social-media-example.png)  
+Let's say you're consuming an API for a social media platform. What kind of data resources are you looking for?
+
+- Posts
+- Pictures
+- Likes
+- Comments
+
+If you're familiar with REST, a REST API endpoints for this app may look like this:
+
+- `/newsfeed` —> [post IDs]
+- `/post/{id}` —> picture ID, user ID, likes
+- `/picture/{id}` —> picture data
+- `/user/{id}/picture` —> commenter profile pic
+
```

Comment:
I don't like the layout of this/how big the image is.  
In notion, they are aligned next to each other. We might want to do that here too. 
<img width="839" alt="image" src="https://user-images.githubusercontent.com/29494270/168412495-a0d4ca68-e5c8-4301-a829-ed4e5a876ddc.png">